readme.md
Description
-----------

**clockdisplay** is a library for seven-segment 4-digits displays, typically used in alarm clocks.
This library is designed for use in Arduino projects.

How to install
--------------

* Copy this folder into arduino sketchs folder\libraries
* Restart the Arduino IDE.
* Enjoy!

How to use
----------

There are a couple of examples, you can see them in your Arduino IDE, in File > Examples > clockdisplay.

Displays
--------

This kind of displays usually have 16 pins to control all the segments:
* Two common cathodes.
* Seven pins for controlling first and second digit segments.
* Seven pins for controlling third and fourth digit segments.

They also have another pins for dots and some extra function, not deemed on this library.

**Supported** display models:
* GC-5062S
* SL-2042-69TB

Connect the displays
--------------------

There is a PDF file in ***models*** folder showing the pinout, so you can find out which pin activates which segment with each cathode, for each model implemented in the library. It also shows how to add additional electronics needed for better performance.


Related libraries
=================

If your display has a cathode (or anode) for each digit and a pin for each segment, then there is a [more suitable library for you](http://code.google.com/p/arduino-seven-segment "Arduino Seven Segment").


history.txt
v0.1 [2015/08/22] (by Miguel C. López)
	+ Original version for GC_6052S and SL_2042_79TB displays.


clockdisplay.h
//
//    FILE: clockdisplay.h
// VERSION: 0.1
// PURPOSE: Control of a clock display
// LICENSE: GPL v3 (http://www.gnu.org/licenses/gpl.html)
//
// WEBPAGE: mclopez.net
//


#ifndef clockdisplay_h
#define clockdisplay_h

#if defined(ARDUINO) && (ARDUINO >= 100)
  #include <Arduino.h>
 #else
  #include <WProgram.h>
#endif

//display models
#define GC_6052S  1
#define SL_2042_79TB	2

class clockdisplay
{
    public:
        clockdisplay(int base_pin, int cathode1_pin, int cathode2_pin, byte display_model);
        void set_d1(bool v);
        void set_d2(bool v);
        void set_d3(bool v);
        void set_d4(bool v);
        void set_byte1(byte b);
        void set_byte2(byte b);
        void set_dec();
        void set_hex();
        void switch_state();
    protected:
        byte bpin, c1, c2;
        byte b1, b2;
        bool v1, v2, v3, v4;
        bool base10;
        byte state;
        byte model;
        void setpins(byte i1, byte i2, byte i3, byte i4);
        
        //for each model, where to get data from pindata
        PROGMEM const byte modeldata =
        	{
        	{2, 0, 1, 0}, //GC_6052S
        	{4, 3, 5, 3}  //SL_2042_79TB
        	};
        			
        //data values for pins
        PROGMEM const byte pindata =
        	{
        	{0xB7, 0x06, 0xE5, 0x67, 0x56, 0x73, 0xF3, 0x07, 0xF7, 0x77, 0xD7, 0xF2, 0xB1, 0xE6, 0xF1, 0xD1, 0x00}, //0
        	{0x7B, 0x30, 0x67, 0x76, 0x3C, 0x5E, 0x5F, 0x70, 0x7F, 0x7E, 0x7D, 0x1F, 0x4B, 0x37, 0x4F, 0x4D, 0x00}, //1
        	{0x00, 0x09, 0x3E, 0x3D, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00}, //2
        	{0x7B, 0x60, 0x5E, 0x76, 0x65, 0x37, 0x3F, 0x70, 0x7F, 0x77, 0x7D, 0x2F, 0x1B, 0x6E, 0x1F, 0x1D, 0x00}, //TODO this is same as pindata[0]
        	{0x00, 0x12, 0x46, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00}, //4
        	{0xBE, 0x06, 0x7C, 0x6E, 0xC6, 0xEA, 0xFA, 0x0E, 0xFE, 0xEE, 0xDE, 0xF2, 0xB8, 0x76, 0xF8, 0xD8, 0x00}  //TODO this is same as pindata[1]
        	};
};

#endif


clockdisplay.cpp
//
//    FILE: clockdisplay.cpp
// VERSION: 0.1
// PURPOSE: Control of a clock display
// LICENSE: GPL v3 (http://www.gnu.org/licenses/gpl.html)
//
// WEBPAGE: mclopez.net
//


#include "clockdisplay.h"


/* =========================
   Constructor
   base_pin:      the Arduino pin connected to the 1st display pin
   cathode1_pin:  the Arduino pin managing the cathode1
   cathode2_pin:  the Arduino pin managing the cathode2
   display_model: the display model
   ========================= */
clockdisplay::clockdisplay(int base_pin, int cathode1_pin, int cathode2_pin, byte display_model)
{
    bpin = base_pin;
    c1 = cathode1_pin;
    c2 = cathode2_pin;
    pinMode(c1, OUTPUT);
    pinMode(c2, OUTPUT);
    for (int i=0; i<14; i++)
    {
        pinMode(bpin+i, OUTPUT);
    }
    state = 0;
    base10 = true;
    v1 = v2 = v3 = v4 = false;
    b1 = b2 = 0;
    model = display_model;
}


/* =========================
   set_dX
   Shows or hides the X digit
   ========================= */
void clockdisplay::set_d1(bool v)
{
    v1 = v;
}

void clockdisplay::set_d2(bool v)
{
    v2 = v;
}

void clockdisplay::set_d3(bool v)
{
    v3 = v;
}

void clockdisplay::set_d4(bool v)
{
    v4 = v;
}


/*
   set_byteX
   Sets the byte1 (digits 1 and 2) or the byte2 (digits 3 and 4)
*/
void clockdisplay::set_byte1(byte b)
{
    b1 = b;
}

void clockdisplay::set_byte2(byte b)
{
    b2 = b;
}


/*
   set_dec
   Show bytes in decimal format
*/
void clockdisplay::set_dec()
{
    base10 = true;
}


/*
   set_hex
   Show bytes in hexadecimal format
*/
void clockdisplay::set_hex()
{
    base10 = false;
}


/*
   swich_state
   This function must be call each 10ms aprox. (between 2ms and 20ms) in order to view the display correctly.
   Call this function only once if you want to turn off the display (after calling set_dX(false)).
*/
void clockdisplay::switch_state()
{
    byte i1, i2, i3, i4;
    i1 = i2 = i3 = i4 = 0x10; //Set all digits as non-visible
    state ^= 1; //switch commons cathodes states
    if (base10) //select decimal digits, if visible
    {
        if (v1)
        {
            i1 = (b1/10)%10;
        }
        if (v2)
        {
            i2 = b1%10;
        }
        if (v3)
        {
            i3 = (b2/10)%10;
        }
        if (v4)
        {
            i4 = b2%10;
        }
    }
    else //select hexadecimal digits, if visible
    {
        if (v1)
        {
            i1 = b1/16;
        }
        if (v2)
        {
            i2 = b1%16;
        }
        if (v3)
        {
            i3 = b2/16;
        }
        if (v4)
        {
            i4 = b2%16;
        }
    }
    setpins(i1, i2, i3, i4);
}


/*
   setpins
   Set the display pins to show tha data
*/
void clockdisplay::setpins(byte i1, byte i2, byte i3, byte i4)
{
    byte s1, s2; //values for display pins

    switch (model)
    {
        case GC_6052S:
            if (s) //set common cathode1 values
            {
                s1 = d2[i2] >> 4;
                s1 |= d1[i1] & 0x70;
                s2 = d2[i4] >> 4;
                s2 |= d3[i3] & 0x70;
            }
            else //set common cathode2 values
            {
                s1 = d2[i2] & 0x07;
                s1 |= d1[i1] << 3;
                s2 = d2[i4] & 0x07;
                s2 |= d3[i3] << 3;
            }
            break;
        case SL_2042_79TB:
            if (s) //set common cathode1 values
            {
                s1 = d2[i2] >> 4;
                s1 |= (d1[i1] & 0x70) >> 1;
                s2 = d2[i4] >> 4;
                s2 |= (d3[i3] & 0xF0) >> 1;
            }
            else //set common cathode2 values
            {
                s1 = d2[i2] & 0x0F;
                s1 |= d1[i1] << 3;
                s2 = d2[i4] & 0x0F;
                s2 |= d3[i3] << 3;
            }
            break;
    }

    //off cathodes to avoid bad view
    digitalWrite(c1,LOW);
    digitalWrite(c2,LOW);

    //set pins
    for (int i=0; i<7; i++)
    {
        digitalWrite(bpin+i, (s2 >> i) & 0x01);
        digitalWrite(bpin+i+7, (s1 >> i) & 0x01);
    }

    //select cathode
    if (state)
    {
        digitalWrite(c2,HIGH);
    }
    else
    {
        digitalWrite(c1,HIGH);
    }
}
